node/bin/src/prover_api/prover_server/v1/handlers.rs

```diff
@@ -99,13 +47,26 @@ async fn submit_fri_proof(
         .decode(&payload.proof)
         .map_err(|e| (StatusCode::BAD_REQUEST, format!("invalid base64: {e}")))?;
 
-    let prover_id = query.id.as_deref().unwrap_or("unknown_prover");
+    let prover_id = query.id;
+    let execution_version = ExecutionVersion::try_from_vk_hash(&payload.vk_hash).map_err(|e| {
+        (
+            StatusCode::BAD_REQUEST,
+            format!("no Execution Version matches the provided verification key: {e}"),
+        )
+    })?;
     match state
         .fri_job_manager
-        .submit_proof(payload.block_number, proof_bytes.into(), prover_id)
+        .submit_proof(payload.block_number, proof_bytes.into(), Some(execution_version), &prover_id)
         .await
     {
         Ok(()) => Ok((StatusCode::NO_CONTENT, "proof accepted".to_string()).into_response()),
+        Err(SubmitError::VerificationKeyHashMismatch(server_vk, prover_vk)) => Err((
```

Comment:
maybe it's worth having execution versions here, not VKs? They are easier to read. We can have both if needed.

Reply:
It's possible that prover_vk has no equivalent for ExecutionVersion (such as `0xblabla`). We can have options, or back port it. I find VKs more obvious (it's the "id" that tells you uniquely what it means), but happy to have both (assuming we do ExecutionVersion as Option<>).